pkg/catalog/actions_source.go

```diff
@@ -4,24 +4,51 @@ import (
 	"context"
 	"fmt"
 	"io"
+	"time"
 
 	"github.com/treeverse/lakefs/pkg/block"
+	"github.com/treeverse/lakefs/pkg/cache"
 	"github.com/treeverse/lakefs/pkg/graveler"
 )
 
 const repositoryLocation = "_lakefs_actions/"
 
 type ActionsSource struct {
 	catalog *Catalog
+	cache   cache.Cache
+}
+
+const (
+	DefaultActionsCacheSize   = 100
+	DefaultActionsCacheExpiry = 5 * time.Second
+	DefaultActionsCacheJitter = DefaultActionsCacheExpiry / 2
+)
+
+type ActionsCacheConfig struct {
+	Size   int
+	Expiry time.Duration
+	Jitter time.Duration
 }
 
 func NewActionsSource(catalog *Catalog) *ActionsSource {
 	return &ActionsSource{
 		catalog: catalog,
+		cache:   cache.NewCache(DefaultActionsCacheSize, DefaultActionsCacheExpiry, cache.NewJitterFn(DefaultActionsCacheJitter)),
 	}
 }
 
 func (s *ActionsSource) List(ctx context.Context, record graveler.HookRecord) ([]string, error) {
+	key := fmt.Sprintf("list_%s_%s", record.RepositoryID.String(), record.SourceRef.String())
+	names, err := s.cache.GetOrSet(key, func() (interface{}, error) {
```

Comment:
Every action except `Commit` is using `commit.ID` which is immutable to load the action files .
We must use the branch to load action files for commit since the action files might have changed and we need to load them and run the latest version. My assumption is that the number of changes to the action files is low compared to changes to data files and with low contention, and a 5 seconds delay introduced by this cache is reasonable.